Connector builder: Tutorial

flash1293 opened the pull request:
## What

Related to https://github.com/airbytehq/airbyte/issues/24953

(not complete, screen recording still missing)

I adapted the existing tutorial for low code connectors - the tutorial can definitely be shortened to get to the "I synced some data" faster, but I think it's nice to keep it a little more realistic. A nice property is that it should work relatively well to skip sections of the tutorial (you don't have to follow everything to a tee to have the next step working).

Once we have the concept pages ready I plan to add links to these from each section so users can diverge from the happy path if they want to.

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* On top of the form, change the "Path URL" input to `/exchangerates_data/{{ stream_slice.start_time }}` to inject the date to fetch data for into the path of the request
+* Open the "Advanced" section and set "Step" to `P1D` to configure the connector to do one separate request per day
+* Set a start date (like `2023-03-03`) in the "Testing values" menu
+* Hit the "Test" button to trigger a new test read
```
I'm getting an error

![Screen Shot 2023-04-14 at 11 01 08 AM](https://user-images.githubusercontent.com/1451943/232122224-6c669506-688c-4883-8861-76ae0e66b751.png)

girarda replied:
Not great, but I was able to fix by formatting the end datetime

![Screen Shot 2023-04-14 at 11 04 24 AM](https://user-images.githubusercontent.com/1451943/232122843-e6dca4f6-da94-4007-a8f7-b2745d25310b.png)

lmossman replied:
I didn't hit any error during this step of the tutorial... not sure why it would be different for us (I'm using CDK version 0.34.1 if that helps). Here is my manifest at this point:
[exchange_rates.yaml.txt](https://github.com/airbytehq/airbyte/files/11237884/exchange_rates.yaml.txt)

flash1293 replied:
You didn't set `today_utc` for the end date @girarda - however, I don't think it matters as we are going to remove this before the release to avoid this kind of error.

flash1293 replied:
Note: We will need to adjust the tutorial a bit once that's done, but I think we don't need to block on anything

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* In a lot of cases the start and end date are injected into the request body or request parameters. However in the case of the exchange rate API it needs to be added to the path of the request, so disable the "Inject start/end time into outgoing HTTP request" options
+* On top of the form, change the "Path URL" input to `/exchangerates_data/{{ stream_slice.start_time }}` to inject the date to fetch data for into the path of the request
+* Open the "Advanced" section and set "Step" to `P1D` to configure the connector to do one separate request per day
+* Set a start date (like `2023-03-03`) in the "Testing values" menu
```
let's make this a more recent date so it's faster. Maybe 2023-04-01?

flash1293 replied:
This is a losing battle - in a month we are back to square one. Also, I don't think a more recent date would make it faster necessarily, this is just the API

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Set a start date (like `2023-03-03`) in the "Testing values" menu
+* Hit the "Test" button to trigger a new test read
+
+Now, you should see a dropdown above the records view that lets you step through the daily exchange rates along with the requests performed to fetch this data. Note that in the connector builder at most 5 slices are requested to speed up testing. During a proper sync the full time range between your configured start date and the current day will be executed.
```
```suggestion
Now, you should see a dropdown above the records view that lets you step through the daily exchange rates along with the requests performed to fetch this data. Note that in the connector builder at most 5 partitions are requested to speed up testing. During a proper sync the full time range between your configured start date and the current day will be executed.
```

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Set the "End datetime" to `{{ today_utc() }}` to always sync up to the current day
+* In a lot of cases the start and end date are injected into the request body or request parameters. However in the case of the exchange rate API it needs to be added to the path of the request, so disable the "Inject start/end time into outgoing HTTP request" options
+* On top of the form, change the "Path URL" input to `/exchangerates_data/{{ stream_slice.start_time }}` to inject the date to fetch data for into the path of the request
+* Open the "Advanced" section and set "Step" to `P1D` to configure the connector to do one separate request per day
```
maybe mention that it'll partition the dataset. 1 partition = 1 day

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+
+## Adding transformations
+
+Note that a warning icon should show next to the "Detected schema" tab - using the per-date endpoint instead of the latest endpoint slightly changed the shape of the records by adding a `historical` property. As we don't need this property in our destination, we can remove it using a transformation.
```
nit: the icon might be a little too subtle. I don't think I would've noticed

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Set the "Path" to `historical`
+* Trigger a new test read
+
+The `historical` property in the records tab and the schema warning should disappear.
```
I didn't see the historical field. Was I supposed to remove the record selector?

lmossman replied:
Yeah I had removed the record selector earlier and I see the historical property.

I ran into a different issue while following this though: I typed the word "historical" into the Path input, and then clicked Test as the tutorial instructed, but the data that was shown still included the `historical` field which was very confusing.

I realized that this was because I never actually "submitted" my changes to the Path array input, so the test ran with the old empty value. I have created an issue for this problem which I'll bring to grooming: https://github.com/airbytehq/airbyte/issues/25228

flash1293 replied:
Eh, record selector should not be set to dates, sorry

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+So far, the connector is only configured as part of the connector builder project. To make it possible to use it in actual connections, you need to publish the connector. This captures the current state of the configuration and makes the connector available as a custom connector within the current Airbyte workspace.
+
+To use the connector for a proper sync, follow these steps:
+* Click the "Publish" button and publish the first version of the "Stock Exchange" connector
```
```suggestion
* Click the "Publish" button and publish the first version of the "Exchange Rates (Tutorial)" connector
```

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+To use the connector for a proper sync, follow these steps:
+* Click the "Publish" button and publish the first version of the "Stock Exchange" connector
+* Go to the "Connections" page and create a new connection
+* As Source type, pick the "Stock Exchange" connector you just created
```
```suggestion
* As Source type, pick the "Exchange Rates (Tutorial)" connector you just created
```

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+## Next steps
+
+This tutorial didn't go into depth about all features that can be used in the connector builder. Check out the concept pages for more information about certain topics:
+* Authentication
```
nit: let's add links to these

flash1293 replied:
yeah, I will do that once everything is merged together (can't point to non-existing pages as there's a link checker testing these)

girarda reviewed (commented):
few comments + made suggestions for screenshots. We can definitely wait until we record the demos though so make sure they're using the latest UX

This is awesome!

girarda commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Click "Set up source" and wait for the connection check to validate the provided configuration is valid
+* Set up a destination - to keep things simple let's choose the "E2E Testing" destination type
+* Click "Set up destination", keeping the default configurations
+* Wait for Airbyte to check the record schema, then click "Set up connection" - this will create the connection and kick off the first sync
```
we should tell the user to change the replication frequency to `Manual`

flash1293 replied:
It didn't seem relevant for the tutorial - that would add two additional steps (setting it to manual and triggering a sync) which can be avoided

lmossman commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+
+The exchange rate API supports configuring a different base currency via request parameter - let's make this part of the user inputs that can be controlled by the user of the connector when configuring a source, similar to the API key.
+
+To do so, follow these steps:
```
The tutorial actually instructs the reader to click on the `Detected schema` tab in the testing side-panel, not the `Declared schema` tab in the stream config view. 

But I did the same thing when going through the tutorial, so we may just want to instruct users to open both view and to switch back to Configuration after importing the schema

lmossman commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Scroll down to the "Request parameters" section and add a new request parameter
+* Set the key to `base`
+* For the value, click the user icon in the input and select "New user input"
+* Set the name to "Base"
```
```suggestion
* Set the name to "Base"
* Click "Create"
```

lmossman commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+
+Now your connector has a second configuration input. To test it, click the "Testing values" button again, set the "Base" to `USD`. Then, click the "Test" button again to issue a new test read.
+
+The record should update to use USD as the base currency:
```
This also resulted in some diffs in my schema, such as:
<img width="463" alt="Screen Shot 2023-04-14 at 5 11 26 PM" src="https://user-images.githubusercontent.com/22731524/232172933-073dbcb2-5e2b-4553-88d2-298b1fb38b60.png">

This could be a good opportunity to explain that in this case, we want to keep the old schema since we want it to be wider, so the warning of the schemas being different should be ignored. This does also add a bit more complexity to the tutorial though so I'm also fine with leaving it out - no strong feelings.

lmossman commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Set the "End datetime" to `{{ today_utc() }}` to always sync up to the current day
+* In a lot of cases the start and end date are injected into the request body or request parameters. However in the case of the exchange rate API it needs to be added to the path of the request, so disable the "Inject start/end time into outgoing HTTP request" options
+* On top of the form, change the "Path URL" input to `/exchangerates_data/{{ stream_slice.start_time }}` to inject the date to fetch data for into the path of the request
+* Open the "Advanced" section and set "Step" to `P1D` to configure the connector to do one separate request per day
```
Agreed that we should set the Step before editing the path since the user is already looking at this part of the UI

lmossman commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+Note that a warning icon should show next to the "Detected schema" tab - using the per-date endpoint instead of the latest endpoint slightly changed the shape of the records by adding a `historical` property. As we don't need this property in our destination, we can remove it using a transformation.
+
+To do so, follow these steps:
+* Enable the "Transformations" section
+* Add a new transformation
```
Not a note for this tutorial doc, but I feel like it would be slightly more intuitive behavior if clicking the checkbox to enable Transformations caused a single empty transformation to be added automatically, rather than also requiring the user to click `+ New transformation` after enabling it. WDYT?

flash1293 replied:
That probably makes sense, could you create an issue / PR for this?

lmossman commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Set up a destination - to keep things simple let's choose the "E2E Testing" destination type
+* Click "Set up destination", keeping the default configurations
+* Wait for Airbyte to check the record schema, then click "Set up connection" - this will create the connection and kick off the first sync
+* After a short while, the sync should complete successfully
```
When I did this, I opened the logs of my job to see the output record and I instead just found a bunch of these deserialization errors:
<img width="2197" alt="Screen Shot 2023-04-14 at 5 39 37 PM" src="https://user-images.githubusercontent.com/22731524/232174593-bc10d061-3fd1-4b69-a1a4-3259fc2b111b.png">

A cursory search on slack revealed this may be because the records are large: https://airbytehq-team.slack.com/archives/C048E3BK1KM/p1680822782177669?thread_ts=1680814742.350359&cid=C048E3BK1KM

It would be nice if users didn't see a bunch of error messages during this tutorial - maybe we can ask the platform team for help here? Only other alternatives I'm imagining are using a different connector with smaller output records (but that would mean rewriting most of this tutorial) or trying to configure the record selector to somehow split up the rates into separate records, but I couldn't figure out an easy way to do that

flash1293 replied:
I don't think this is related to the size of the record, our records are fairly small.

I think it's a bug of the e2e destination - the platform tries to interpret its log messages as airbyte messages:
```
2023-04-17 11:10:13 ERROR i.a.w.i.VersionedAirbyteStreamFactory(toAirbyteMessage):360 - Deserialization failed: "2023-04-17 11:10:13 \u001B[32mINFO\u001B[m i.a.i.d.e.TestingDestinations(main):71 - starting destination: class io.airbyte.integrations.destination.e2e_test.TestingDestinations"
```

You can see that the log message `starting destination: class io.airbyte.integrations.destination.e2e_test.TestingDestinations` gets parsed as an Airbyte message which fails. I'm not sure who can help, I will ask the destinations team.

I think we should fix this bug in the e2e connector rather than switching away from it as there is no destination that works without configuration which keeps the tutorial focused.

lmossman reviewed (commented):
Left a bunch of small comments, but overall this tutorial is great and easy to follow!

flash1293 commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Scroll down to the "Request parameters" section and add a new request parameter
+* Set the key to `base`
+* For the value, click the user icon in the input and select "New user input"
+* Set the name to "Base"
```
I originally had that but removed it as it's not important (it's fairly easy to understand) and I want to keep it focused. I think we can keep it in the screenshots / the video walkthrough though

flash1293 commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Set the name to "Rates"
+* Set the "URL path" to `/exchangerates_data/latest`
+* Submit
+* Set the "Record selector" to `rates`
```
I wanted to keep this out of the tutorial as it would require you to go into detail of what a primary key is doing. Let's leave that for the "Record selection" page

flash1293 commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+
+This form corresponds to what a user of this connector will need to provide when setting up a connection later on. The testing values are not saved along with the connector project.
+
+Now, click the "Test" button to trigger a test read to simulate what will happen during a sync. After a little while, you should see a single record that looks like this:
```
yeah, let's leave it out

flash1293 commented on docs/connector-development/connector-builder-ui/tutorial.md:
```diff
@@ -0,0 +1,202 @@
+* Set the name to "Rates"
+* Set the "URL path" to `/exchangerates_data/latest`
+* Submit
+* Set the "Record selector" to `rates`
```
agreed, not necessary for the scope of the tutorial

flash1293 commented:
Thanks for the review @lmossman and @girarda . I think I addressed all your points. I added some images for the parts of the UI that won't change anymore (which excludes the incremental sync section) - however I created new images with a higher resolution so the screenshots render sharp even on retina displays.

I didn't capture every step via screenshot because I will add a small recording to every step that should fulfil the same use case, but it's still a nice touch to have the screenshots.

flash1293 commented on docs/connector-development/connector-builder-ui/tutorial.mdx:
```diff
@@ -0,0 +1,226 @@
+import NewStreamModal from './assets/tutorial-new-stream.png';
```
yes, I need to use this syntax to be able to add an image and enforce a max width...
